scenarios/smart_trailer/applications/smart_trailer_application/src/main.rs

```diff
@@ -0,0 +1,252 @@
+const CHARIOTT_SERVICE_DISCOVERY_URI: &str = "http://0.0.0.0:50000";
+
+const DEFAULT_FREQUENCY_MS: u64 = 10000; // 10 seconds
+
+// Constants used for retry logic
+const MAX_RETRIES: i32 = 10; // for demo purposes we will retry a maximum of 10 times
+                             // By default we will wait 5 seconds between retry attempts
+const DURATION_BETWEEN_ATTEMPTS: Duration = Duration::from_secs(5);
+
+/// Get trailer weight's subscription information from managed subscribe endpoint.
+///
+/// # Arguments
+/// * `managed_subscribe_uri` - The managed subscribe URI.
+/// * `constraints` - Constraints for the managed topic.
+async fn get_trailer_weight_subscription_info(
+    managed_subscribe_uri: &str,
+    constraints: Vec<Constraint>,
+) -> Result<SubscriptionInfoResponse, Status> {
+    // Create gRPC client.
+    let mut client = ManagedSubscribeClient::connect(managed_subscribe_uri.to_string())
+        .await
+        .map_err(|err| Status::from_error(err.into()))?;
+
+    let request = Request::new(SubscriptionInfoRequest {
+        entity_id: trailer_v1::trailer::trailer_weight::ID.to_string(),
+        constraints,
+    });
+
+    let response = client.get_subscription_info(request).await?;
+
+    Ok(response.into_inner())
+}
+
+/// Receive Trailer Weight updates.
+///
+/// # Arguments
+/// * `broker_uri` - The broker URI.
+/// * `topic` - The topic.
+async fn receive_trailer_weight_updates(
+    broker_uri: &str,
+    topic: &str,
+) -> Result<JoinHandle<()>, String> {
+    // Create a unique id for the client.
+    let client_id = format!("{MQTT_CLIENT_ID}-{}", Uuid::new_v4());
+
+    let create_opts = mqtt::CreateOptionsBuilder::new()
+        .server_uri(broker_uri)
+        .client_id(client_id)
+        .finalize();
+
+    let client = mqtt::Client::new(create_opts)
+        .map_err(|err| format!("Failed to create the client due to '{err:?}'"))?;
+
+    let receiver = client.start_consuming();
+
+    // Setup task to handle clean shutdown.
+    let ctrlc_cli = client.clone();
+    tokio::spawn(async move {
+        _ = signal::ctrl_c().await;
+
+        // Tells the client to shutdown consuming thread.
+        ctrlc_cli.stop_consuming();
+    });
+
+    // Last Will and Testament
+    let lwt = mqtt::MessageBuilder::new()
+        .topic("test")
+        .payload("Receiver lost connection")
+        .finalize();
+
+    let conn_opts = mqtt::ConnectOptionsBuilder::new_v5()
+        .keep_alive_interval(Duration::from_secs(30))
+        .clean_session(false)
+        .will_message(lwt)
+        .finalize();
+
+    client
+        .connect(conn_opts)
+        .map_err(|err| format!("Failed to connect due to '{err:?}"))?;
+
+    client
+        .subscribe(topic, mqtt::types::QOS_1)
+        .map_err(|err| format!("Failed to subscribe to topic {topic} due to '{err:?}'"))?;
+
+    // Copy topic for separate thread.
+    let topic_string = topic.to_string();
+
+    let sub_handle = tokio::spawn(async move {
```

Comment:
To make use of Will's suggestion change this line to `let sub_handle: JoinHandle<Result<(), String>> = tokio::spawn(async move {`